Docs: Add routing.md

ockham opened the pull request:
Are there any gaping holes in the logical structure of this doc? Is it useful?

seear commented:
We can add that a huge drawback of having both multi- and single-tree rendering is that the entire layout needs to be re-rendered when switching between them.

aduth commented:
Do we not want to touch on the variation of sections which accept the router as an argument?

e.g. https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/types/index.js

I was under the impression this was "preferred", even in client-side-only routing.

ockham commented:
@aduth Good point. The hair in the soup is that this also does the rendering, which I myself have found a bit too magical in the past; it also prevents us from using that kind of [separating concerns](https://visionmedia.github.io/page.js/#separating-concerns) that `page.js` would otherwise allow. (The alternative I'm ensvisioning would be to pass not only a `router` arg but also a `render` one.)

(It's also covered in `docs/isomorphic-routing.md`.)

> I was under the impression this was "preferred", even in client-side-only routing.

I guess the question is, how much is this doc descriptive (from which new devs trying to grasp Calypso will no doubt benefit) or normative (recommending best practices for future use) -- in a similar vein as your `our-approach-to-data.md`.

Thinking about this some more, I guess you're right in that we should ask devs that want to write a new section to use `router`, since it paves the way to single-tree rendering those sections. I'll add something.

aduth commented:
>The hair in the soup is that this also does the rendering, which I myself have found a bit too magical in the past

Personally I find this to be the perfectly acceptable amount of magic. Maybe I'm biased from traumatic experience of having done the `React.render` -> `ReactDom.render` migration (#1498) 😆 .

ockham commented:
> We can add that a huge drawback of having both multi- and single-tree rendering is that the entire layout needs to be re-rendered when switching between them.

@seear I already had sentence that we have to unmount, I added the re-rendering part to it now -- I hope it gets across strongly enough: d136c6c

@aduth Added the `router` arg recommendation. I also added a note about the unmounting/rerendering `page()` handler in `client/boot` and the STR whitelist to which authors of STR'ed sections have to add theirs.

seear commented:
> @seear I already had sentence that we have to unmount, I added the re-rendering part to it now -- I hope it gets across strongly enough: d136c6c

Ah, no, sorry, I mean we can say that _the whole app is re-rendered_, which is obviously non-optimal and causers a screen flicker. Just to emphasize why the current situation is not really acceptable.

ockham commented:
@seear Ah, gotcha. 30ce2ca then 😄
